src/lib/ikea.ts

```diff
@@ -0,0 +1,662 @@
+    ];
+
+    const fromZigbee: Fz.Converter[] = [{
+        cluster: 'genPowerCfg',
+        type: ['attributeReport', 'readResponse'],
+        convert: (model, msg, publish, options, meta) => {
+            const payload: KeyValue = {};
+            if (msg.data.hasOwnProperty('batteryPercentageRemaining') && (msg.data['batteryPercentageRemaining'] < 255)) {
+                // Some devices do not comply to the ZCL and report a
+                // batteryPercentageRemaining of 100 when the battery is full (should be 200).
+                //
+                // IKEA corrected this on newer remote fw version, but many people are still
+                // 2.2.010 which is the last version supporting group bindings. We try to be
+                // smart and pick the correct one for IKEA remotes.
+                let dividePercentage = true;
+                // If softwareBuildID is below 2.4.0 it should not be divided
+                if (semver.lt(meta.device.softwareBuildID, '2.4.0', true)) {
+                    dividePercentage = false;
+                }
+                let percentage = msg.data['batteryPercentageRemaining'];
+                percentage = dividePercentage ? percentage / 2 : percentage;
+                payload.battery = precisionRound(percentage, 2);
+            }
+
+            return payload;
+        },
+    }];
+
+    const toZigbee: Tz.Converter[] = [{
+        key: ['battery'],
+        convertGet: async (entity, key, meta) => {
+            await entity.read('genPowerCfg', ['batteryPercentageRemaining']);
+        },
+    }];
+
+    const defaultReporting: ReportingConfigWithoutAttribute = {min: '1_HOUR', max: 'MAX', change: 10};
+
+    const configure: Configure = async (device, coordinatorEndpoint, logger) => {
+        await setupAttributes(device, coordinatorEndpoint, 'genPowerCfg', [
+            {attribute: 'batteryPercentageRemaining', ...defaultReporting},
+        ], logger);
+    };
+
+    return {exposes, fromZigbee, toZigbee, configure, isModernExtend: true};
+}
+
+export function ikeaConfigureRemote(): ModernExtend {
+    const configure: Configure = async (device, coordinatorEndpoint, logger) => {
+        // Firmware 2.3.075 >= only supports binding to endpoint, before only to group
+        // - https://github.com/Koenkk/zigbee2mqtt/issues/2772#issuecomment-577389281
+        // - https://github.com/Koenkk/zigbee2mqtt/issues/7716
+        const endpoint = device.getEndpoint(1);
+        const version = device.softwareBuildID.split('.').map((n) => Number(n));
+        const bindTarget = version[0] > 2 || (version[0] == 2 && version[1] > 3) || (version[0] == 2 && version[1] == 3 && version[2] >= 75) ?
+            coordinatorEndpoint : constants.defaultBindGroup;
+        await endpoint.bind('genOnOff', bindTarget);
+    };
+
+    return {configure, isModernExtend: true};
+}
+
+export function ikeaAirPurifier(): ModernExtend {
```

Comment:
It's probably more readable to split these into multiple modernExtends e.g.
- ikeaAirPurifierPm25
- ikeaAirPurifierFilterRunTime
- ikeaAirPurifierControlPanelLight
- ikeaAirPurifierChildLock
- ikeaAirPurifierFanSpeed
- ikeaAirPurifierFanMode

Each just having the expose, fromZigbee, toZigbee and configure that is relavent for the attribute.

Reply:
Is it necessary though? It’s not like this is going to be reused in parts anyway. Everything here comes from the same ikea purifier specific cluster

Reply:
Functionally it won't make much difference, just readability.

I'll probably eventually do it myself, the current bug blob of code is not very readable. The person who originally wrote it did not do a very good job at keeping it readable. (I wrote it so It's OK to say it without offending anyone)